peps/pep-0750.rst

```diff
@@ -53,837 +33,1089 @@ deferred pending further experience with f-strings. Work on this PEP was
-is a string-like object with extra powers, as described below. An "interpolation" object is a
-tuple-like value representing how Python processed the interpolation into a form useful for your
-tag function. Both are fully described below in `Specification`_.
+This PEP introduces a new string prefix, ``t``, to define template string literals.
+These literals resolve to a new type, ``types.Template``.
 
-Here is a more generalized version using structural pattern matching and type hints:
+The following code creates a ``Template`` instance:
 
 .. code-block:: python
 
-    from typing import Decoded, Interpolation  # Get the new protocols
+    from types import Template
+    template = t"This is a template string."
+    assert isinstance(template, Template)
 
-    def greet(*args: Decoded | Interpolation) -> str:
-        """Handle arbitrary args using structural pattern matching."""
-        result = []
-        for arg in args:
-            match arg:
-                case Decoded() as decoded:
-                    result.append(decoded)
-                case Interpolation() as interpolation:
-                    value = interpolation.getvalue()
-                    result.append(value.upper())
+Template string literals support the full syntax of :pep:`701`. This includes
+the ability to nest template strings within interpolations, as well as the ability
+to use any quote mark in the string literal. Like other string prefixes, the template
+prefix must immediately precede the quote mark. Like f-strings, both lowercase 
+``t`` and uppercase ``T`` prefixes are supported. Like f-strings, t-strings may
+not be combined with the ``b`` or ``u`` prefixes. They *may* be combined with 
+the ``r`` prefix; see the `Raw Template Strings`_ section below for more information.
 
-        return f"{''.join(result)}!"
 
-    name = "World"
-    greeting = greet"Hello {name} nice to meet you"
-    assert greeting == "Hello WORLD nice to meet you!"
+The ``Template`` Type
+---------------------
 
-Tag strings extract more than just a callable from the ``Interpolation``. They also
-provide Python string formatting info, as well as the original text:
+Template strings evaluate to an instance of a new type, ``Template``, found
+in the ``<TBC>`` module (proposed: ``types``):
 
 .. code-block:: python
 
-    def greet(*args: Decoded | Interpolation) -> str:
-        """Interpolations can have string formatting specs and conversions."""
-        result = []
-        for arg in args:
-            match arg:
-                case Decoded() as decoded:
-                    result.append(decoded)
-                case getvalue, raw, conversion, format_spec:  # Unpack
-                    gv = f"gv: {getvalue()}"
-                    r = f"r: {raw}"
-                    c = f"c: {conversion}"
-                    f = f"f: {format_spec}"
-                    result.append(", ".join([gv, r, c, f]))
+    @dataclass(frozen=True)
```

Comment:
Will this be a dataclass at runtime (`assert dataclasses.is_dataclass(t'') is True`)?

Reply:
(Note I'm not sure it should be, but if it isn't, we shouldn't advertise that it will be)

Reply:
From the PEP:

> The use of ``@dataclass`` in the definition of ``Template`` above (and 
``Interpolation`` below) is meant to be suggestive; the exact implementation 
in CPython may differ, but developers can expect that ``Template`` instances 
can be constructed and utilized in the same way as a typical dataclass.

We went back and forth on this during editing.

@lysnikolaou Maybe I should just bite the bullet, write an `__init__()` for both `Template` and `Interpolation`, and I suppose also indicate their `__match_args__`?

Reply:
I assume Template will be written in C and therefore cannot be exactly a dataclass.

Reply:
I'd support the ``__init__`` approach: better to be explicit, and an extra few lines won't meaningfully affect the length of this PEP.

Reply:
@AA-Turner @lysnikolaou Done!